server/src/main/java/org/elasticsearch/index/analysis/ReloadableCustomAnalyzer.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.analysis;
+
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.util.CloseableThreadLocal;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.analysis.CustomAnalyzerProvider.AnalyzerComponents;
+
+import java.io.Reader;
+import java.util.Map;
+
+public class ReloadableCustomAnalyzer extends Analyzer implements TokenFilterComposite {
+
+    private volatile AnalyzerComponents components;
+
+    private CloseableThreadLocal<AnalyzerComponents> storedComponents = new CloseableThreadLocal<>();
+
+    private final int positionIncrementGap;
+
+    private final int offsetGap;
+
+    private static final ReuseStrategy UPDATE_STRATEGY = new ReuseStrategy() {
+        @Override
+        public TokenStreamComponents getReusableComponents(Analyzer analyzer, String fieldName) {
+            ReloadableCustomAnalyzer custom = (ReloadableCustomAnalyzer) analyzer;
+            AnalyzerComponents components = custom.getStoredComponents();
+            if (components == null || custom.shouldReload(components)) {
+                custom.setStoredComponents(custom.getComponents());
+                return null;
+            }
+            TokenStreamComponents tokenStream = (TokenStreamComponents) getStoredValue(analyzer);
+            assert tokenStream != null;
+            return tokenStream;
+        }
+
+        @Override
+        public void setReusableComponents(Analyzer analyzer, String fieldName, TokenStreamComponents tokenStream) {
+            setStoredValue(analyzer, tokenStream);
+        }
+    };
+
+    ReloadableCustomAnalyzer(AnalyzerComponents components, int positionIncrementGap, int offsetGap) {
+        super(UPDATE_STRATEGY);
+        this.components = components;
+        this.positionIncrementGap = positionIncrementGap;
+        this.offsetGap = offsetGap;
+    }
+
+    public AnalyzerComponents getComponents() {
+        return this.components;
+    }
+
+    @Override
+    public TokenFilterFactory[] tokenFilters() {
+        return this.components.getTokenFilters();
```

Comment:
This is dangerous since we shouldn't use these filters if we also need to access other members of the `AnalyzerComponents`. Why do you need this `TokenFilterComposite`? It seems that it is used only in one place so I wonder if we could remove this function and adapt the code to check for `CustomAnalyzer` or `ReloadableCustomAnalyzer` ?

Reply:
I introduced the TokenFilterComposite in three places (NamedAnalyzer.checkAllowedInMode, FragmentBuilderHelper.containsBrokenAnalysis, PhraseSuggestionBuilder.getShingleFilterFactory) and one test where we otherwise would have to do perform two `instanceof` checks and have different access patterns for the `CustomAnalyzer` or `ReloadableCustomAnalyzer` case. In both cases what we really only need read access to the filters to check some of their properties. 

> This is dangerous since we shouldn't use these filters if we also need to access other members of the AnalyzerComponents

I understand, but in these cases we are only accessing the filters, no other components (and only on read), so I don't think its dangerous and simplifies the code. I'd like to keep it to keep the code simple, wdyt?